### ISSUE TITLE
decode the frame at the given position

### ISSUE THREAD

**Issue by 756890911** [NVIDIA/VideoProcessingFramework#303]
I want to decode the frame at the given position directly, but hint
`
[matroska,webm @ 0x55614c308500] File is broken, keyframes not correctly marked!
`

Here is my code. 
```python
packet = np.ndarray(shape=(0),dtype = np.uint8)
seek_ctx = nvc.SeekContext(count, nvc.SeekMode.EXACT_FRAME, nvc.SeekCriteria.BY_NUMBER)
nvDmx.Seek(seek_ctx,packet)

pdata_in= nvc.PacketData()
nvDmx.LastPacketData(pdata_in)

surface_nv12 = nvDec.DecodeSurfaceFromPacket(pdata_in,packet)
```

**Comment by 756890911** [comment]
I find that surface_nv12.Empty() is true.

**Comment by rarzumanyan** [comment]
Hi @756890911 

You're seeking for exact frame with demuxer which means it will search for particular encoded frame. Demuxer doesn't decode anything so it will basically find encoded packet with desired timestamp.

There's no guarantee decoder can restore actual pixels from this frame. E. g. because the frame you're seeking for is a P or B frame and it uses other frames (which aren't decoded yet) for prediction.

I recommend you to seek with decoder first as it's shown here:
https://github.com/NVIDIA/VideoProcessingFramework/blob/ec453adcb2c3cd5cd566300d613d3ffd63343dda/Tests/test_PyNvDecoder.py#L226-L232

**Comment by 756890911** [comment]
Thanks for your quick reply. 
Decoder can't accurately locate a frame, only find the most recent I frame.
seek_ctx = nvc.SeekContext(count, nvc.SeekMode.EXACT_FRAME, nvc.SeekCriteria.BY_NUMBER)
RuntimeError: Decoder can only seek to closest previous key frame

**Comment by rarzumanyan** [comment]
@756890911 

>Decoder can't accurately locate a frame, only find the most recent I frame.

That's how seek works in a nutshell.

E. g. you're searching for frame №27. Closest previous I frame is №24. Decoder will jump to it, decode 3 more frames internally will return you reconstructed frame №27.

**Comment by 756890911** [comment]
I tried it
seek_ctx = nvc.SeekContext(count, nvc.SeekMode.PREV_KEY_FRAME, nvc.SeekCriteria.BY_NUMBER)

For example: count = 88

Compared with the continuous decoding frame 88, it is not the same picture.

So instead of returning a frame at 88, it's a previous frame.

**Comment by rarzumanyan** [comment]
@756890911 

Before delving too deep: did you start count from 0 in both cases?

**Comment by 756890911** [comment]
First of all, I apologize for the description.

The frame decoded by count=88 is the same as the frame decoded continuously at the position around 30.

**Comment by rarzumanyan** [comment]
Ok, so I assume that decoder seeks for a frame №30 which is key frame and doesn't go any further.
Taking into account that your input mkv file has some issues:
`[matroska,webm @ 0x55614c308500] File is broken, keyframes not correctly marked!`

Does it reproduce on other videos?

**Comment by 756890911** [comment]
Yes, I tried the new.mkv file again, the above error occurred.
[matroska,webm @ 0x55614c308500] File is broken, keyframes not correctly marked!
But there is no problem using.mp4 files.

At the same time
surface_nv12 = nvDec.DecodeSurfaceFromPacket(pdata_in,packet)
Surface_nv12 is always empty.

**Comment by 756890911** [comment]
A new experiment has shown that decoder seeks for a frame which is key frame and doesn't go any further. 
seek_ctx = nvc.SeekContext(count, nvc.SeekCriteria.BY_NUMBER)
rawSurface = nvDec.DecodeSingleSurface(seek_ctx)

**Comment by rarzumanyan** [comment]
Can you share your mkv file?
I'm under impression that this undesired VPF behavior is specific to the problematic input.

**Comment by 756890911** [comment]
I have sent an email.

**Comment by rarzumanyan** [comment]
Thanks for the update, @756890911 

I've used your input file for `test_PyNvDecoder.py` unit test and for some frame numbers it seeks correctly, for some not:
```bash
roman@HP-ZBook-Studio-G7:~/Git/VideoProcessingFramework/install/bin$ python3 ./test_PyNvDecoder.py
................
----------------------------------------------------------------------
Ran 16 tests in 1.449s

OK
roman@HP-ZBook-Studio-G7:~/Git/VideoProcessingFramework/install/bin$ python3 ./test_PyNvDecoder.py
..........F.F...
======================================================================
FAIL: test_decodesinglesurface_cmp_vs_continuous (__main__.TestDecoderBuiltin)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./test_PyNvDecoder.py", line 269, in test_decodesinglesurface_cmp_vs_continuous
    self.fail('Video frames are not equal')
AssertionError: Video frames are not equal

======================================================================
FAIL: test_decodesinglesurface_seek (__main__.TestDecoderBuiltin)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./test_PyNvDecoder.py", line 238, in test_decodesinglesurface_seek
    self.assertEqual(gt_num_frames-start_frame, dec_frames)
AssertionError: 8 != 9

----------------------------------------------------------------------
Ran 16 tests in 1.410s

FAILED (failures=2)
```

These are test which sometimes passes and sometimes not:
```python
    def test_decodesinglesurface_seek(self):
        start_frame = random.randint(0, gt_num_frames-1)
        dec_frames = 1
        seek_ctx = nvc.SeekContext(
            seek_frame=start_frame, seek_criteria=nvc.SeekCriteria.BY_NUMBER)
        surf = self.nvDec.DecodeSingleSurface(seek_ctx)
        self.assertNotEqual(True, surf.Empty())
        while True:
            surf = self.nvDec.DecodeSingleSurface()
            if surf.Empty():
                break
            dec_frames += 1
        self.assertEqual(gt_num_frames-start_frame, dec_frames)

    def test_decodesinglesurface_cmp_vs_continuous(self):
        # First get reconstructed frame with seek
        start_frame = random.randint(0, gt_num_frames-1)
        seek_ctx = nvc.SeekContext(
            seek_frame=start_frame, seek_criteria=nvc.SeekCriteria.BY_NUMBER)
        seek_frame = np.ndarray(shape=(0), dtype=np.uint8)
        self.assertTrue(self.nvDec.DecodeSingleFrame(seek_frame, seek_ctx))

        # Then get it with continuous decoding
        self.nvDec = nvc.PyNvDecoder(gt_file, 0)
        curr_frame = np.ndarray(shape=(0), dtype=np.uint8)
        curr_frame_id = 0
        while curr_frame_id != start_frame:
            self.assertTrue(self.nvDec.DecodeSingleFrame(curr_frame))
            curr_frame_id += 1

        # Compare frames
        if not np.array_equal(seek_frame, curr_frame):
            # First dump them on disk to inspect visually
            seek_frame_handle = open('./seek_frame.yuv', 'wb')
            bits = bytearray(seek_frame)
            seek_frame_handle.write(bits)
            seek_frame_handle.close()

            curr_frame_handle = open('./curr_frame.yuv', 'wb')
            bits = bytearray(curr_frame)
            curr_frame_handle.write(bits)
            curr_frame_handle.close()

            self.fail('Video frames are not equal')
```

I tried to convert from mkv container to mpg but ffmpeg fails to do so without re-encoding:
![image](https://user-images.githubusercontent.com/45193474/150182032-2a92d538-dd33-4601-af3b-39f992b6eba5.png)


However, the original mkv file structure looks OK in bitstream analyzer:
![image](https://user-images.githubusercontent.com/45193474/150181424-7221f652-74ad-487e-b197-bd798e51491f.png)

I'm under impression that something's happening under the hood of FFMpeg API which is used for seeking. Will try to reproduce on different files and / or after re-encoding.

**Comment by rarzumanyan** [comment]
@756890911 

A quick update:
The input video has some inconsistency between DTS values and duration.
Please find below the output of `ffprobe -show_packets ./out.mkv`:
```bash
[PACKET]
codec_type=video
stream_index=0
pts=29
pts_time=0.029000
dts=29
dts_time=0.029000
duration=28
duration_time=0.028000
convergence_duration=N/A
convergence_duration_time=N/A
size=12879
pos=127577
flags=__
[/PACKET]
[PACKET]
codec_type=video
stream_index=0
pts=86
pts_time=0.086000
dts=57
dts_time=0.057000
duration=28
duration_time=0.028000
convergence_duration=N/A
convergence_duration_time=N/A
size=12260
pos=140463
flags=__
[/PACKET]
[PACKET]
codec_type=video
stream_index=0
pts=229
pts_time=0.229000
dts=86
dts_time=0.086000
duration=28
duration_time=0.028000
convergence_duration=N/A
convergence_duration_time=N/A
size=46898
pos=152731
flags=__
[/PACKET]
```

All the packets in video has same duration of `28`. However, DTS values aren't always of duration. E. g. the sequence of DTS values is: `29, 57, 86`. As you see, `57==29+28`, however `86!=57+28`. DTS and duration values are used to convert packet timestamp values to frame number and this may cause off-by-one errors which happen from time to time.

I'm not an expert in this field and can't tell from the top of my head if DTS must always be multiple of duration, so I'll need some time to investigate on this. At least now we know what's causing off-by-one errors in seek.

**Comment by 756890911** [comment]
Thank you for your answer. I will also conduct research on this part and update it here if there is any latest progress.

**Comment by 756890911** [comment]
I found that there was a problem inside the data and changed the data. This is a new difficulty at present. Surface_nv12 is empty and I find pdata_in.dts is -9223372036854775808. Here is my code. 

seek_ctx = nvc.SeekContext(count ,nvc.SeekMode.PREV_KEY_FRAME, nvc.SeekCriteria.BY_NUMBER)
packet = np.ndarray(shape=(0),dtype = np.uint8)
pdata_in = nvc.PacketData()

nvDmx.Seek(seek_ctx,packet)

print(pdata_in.duration)
print(pdata_in.dts)
nvDmx.LastPacketData(pdata_in)
print(pdata_in.duration)
print(pdata_in.dts)

surface_nv12 = nvDec.DecodeSurfaceFromPacket(pdata_in,packet)

**Comment by rarzumanyan** [comment]
Hi @756890911 

I don't fully understand why do you use demuxer for seek. 
It's much easier to do with decoder as it's illustrated above. Could you try that?

**Comment by 756890911** [comment]
I have communicated with you about this before, because the decoder is not used to find the frame in the specified position, but the nearest key frame before it. Then I tried demuxer.

**Comment by 756890911** [comment]
Maybe I don't understand the tool properly. I have implemented continuous reading and decoding of frames. Now I want to skip some of the frames and decode them for a particular location.

**Comment by rarzumanyan** [comment]
@756890911 

>I have communicated with you about this before, because the decoder is not used to find the frame in the specified position, but the nearest key frame before it.

Probably there's some disconnect here, let me explain:

- Both demuxer and decoder classes in VPF can be used to seek, but their behavior is different.
- When you seek with decoder, it will go to the closest I frame back in the video because it needs that key frame to reconstruct the frame you are seeking for. Frames within the GOP (Group Of Pictures) use other frames within same GOP for reference to improve compression. Hence you can't just decode random frame, you have to start with I frame which is in the start of GOP.
- Demuxer can seek to any frame because it doesn't have to decode any pictures. It just goes through video file and parses the headers of the frames. It can be used e. g. if you want to extract video stream from container to another file without re-encoding.

Long story short: I recommend you to seek with decoder unless you're 100% sure you need demuxer.
If you like to get more details, [this](https://en.wikipedia.org/wiki/Group_of_pictures) Wiki article may be a good place to start.

**Comment by 756890911** [comment]
When I find the most recent I frame in the video, is there a way to proceed and make it decode the frame I was trying to decode?

**Comment by rarzumanyan** [comment]
>When I find the most recent I frame in the video, is there a way to proceed and make it decode the frame I was trying to decode?

Decoder will do this for you.
Search for closest I frame and restoration of reference frames happens under the hood.
Code sample above shall return you the exact frame you're seeking for (this is what unit test actually checks).

**Comment by 756890911** [comment]
Thanks a lot.

**Comment by rarzumanyan** [comment]
Hi @756890911 

I'm reopening this because I've been working on more accurate seek in the background.
Please check out `seek_cleanup` branch, it allows for more accurate seek based on fps instead of duration.

I've checked your original input file seeking for every frame from 0 to 106.
All frames obtained with seek are bit-to-bit same as that obtained with continuous decoding except the last one. 
It has weird `PTS=3086` value which doesn't correspond to it's number divided by FPS:

```bash
Traceback (most recent call last):
  File "./test_PyNvDecoder.py", line 283, in test_decodesinglesurface_cmp_vs_continuous
    self.fail(fail_msg)
AssertionError: Seek frame number: 106.
Seek frame pts:    3029.
Cont frame pts:    3086.
Video frames are not same
```

Frame №106 shall have `PTS=106/35*1000=3028.57` which rounds up to `3029`, but in the bitstream it has `PTS=3086` for some reason.

**Comment by rarzumanyan** [comment]
Closing as fixed.
